### PR TITLE
Expose interface that represents unencrypted tokens

### LINKED ISSUE #203: How get Claims after encode token
I'm try to get claim on middleware.
Using Version 4.0@dev
image
But `parse` function return Plain, this is a error ?
image

### LINKED ISSUE #228: Getting claims from Token
Hi, I noticed that the `Plain` token has a [`claims` method](https://github.com/lcobucci/jwt/blob/master/src/Token/Plain.php#L49), but it's not part of the `Token` interface.

Is this missing on purpose? If not I could create PR to add it if you like.

```php
interface Token {
    public function claims(): DataSet;
}
```

### LINKED ISSUE #611: Should the TokenInterface contain a method for claims() ?
After upgrading to v4 of the library my static code analysis is failing because claims() is not defined Token interface. I guess this method should be declared, also because the code of the library itself is depending on it but others as well. Do you agree? If so I will create a PR for this change.

### PR DESCRIPTION
**Pull request by Slamdunk**
Resolve https://github.com/lcobucci/jwt/pull/186
Resolve https://github.com/lcobucci/jwt/issues/228
Resolve https://github.com/lcobucci/jwt/issues/203
Resolve https://github.com/lcobucci/jwt/pull/552
Resolve https://github.com/lcobucci/jwt/issues/611
Resolve https://github.com/lcobucci/jwt/pull/615

### CODE REVIEW

File: src/PlainToken.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?php
+declare(strict_types=1);
+
+namespace Lcobucci\JWT;
+
+use Lcobucci\JWT\Token\DataSet;
+
+interface PlainToken extends Token
```
Comment (lcobucci) [inline src/PlainToken.php]: I'd call this `UnencryptedToken` in this context.

**Review by lcobucci — CHANGES_REQUESTED** [review]
We also need to change the SignedWith constraint: https://github.com/lcobucci/jwt/blob/4.1.x/src/Validation/Constraint/SignedWith.php#L24

(and possibly need to add the payload method to the this interface as well).

**Comment by Slamdunk** [comment]
> (and possibly need to add the payload method to the this interface as well).

I didn't change `SignedWith` because indeed it also requires `payload` and `signature`, two methods only useful within the `SignedWith` context.
Just one last confirmation: do you want me to add the two methods mentioned in the new `UnencryptedToken` interface?

**Comment by lcobucci** [comment]
> > (and possibly need to add the payload method to the this interface as well).
> 
> I didn't change `SignedWith` because indeed it also requires `payload` and `signature`, two methods only useful within the `SignedWith` context.
> Just one last confirmation: do you want me to add the two methods mentioned in the new `UnencryptedToken` interface?

Yes, please. These are relevant to unencrypted tokens.

The main benefit we gain here is to allow users to provider their own builders/parsers/tokens implementation and everything will just work.

**Review by lcobucci — APPROVED** [review]
The only thing we're missing to allow people to make full use of this is to change the builder to use the interface as well... That's a minor BC-break, though. I'll think about it for the next few days. Let's get this merged, though.

@Slamdunk thanks!

**Comment by lcobucci** [comment]
@Slamdunk would please sign your commits? Otherwise I will rebase your branch on the weekend.

**Comment by Slamdunk** [comment]
The users don't **need** `\Lcobucci\JWT\Builder::getToken` to get changed, even though of course it would be good in the next major. It already return a `Plain` token.

Users will still need `assert($token instanceof UnencryptedToken);` because the place where they need this is after `\Lcobucci\JWT\Parser::parse` call, which should always return a `\Lcobucci\JWT\Token` in case encrypted token will be implemented, afaict.

> would please sign your commits?

Yeah, sorry, It's weird to me that commits appear unsigned just because the key is expired. I'm waiting my Yubikeys, in the meantime I'm using short-lived keys.

I'll re-sign soon

**Comment by lcobucci** [comment]
@Slamdunk thanks!

**Pull request merged**